provider/pkg/provider/ids.go

```diff
@@ -53,14 +93,24 @@ func getDefaultName(
 		randLength = autoNamingSpec.MinLength - len(prefix)
 	}
 
+	// Generate naming trivia for the resource.
+	namingTrivia := createNamingTrivia(sdkName, news)
+
 	maxLength := 0
 	if autoNamingSpec.MaxLength > 0 {
-		left := autoNamingSpec.MaxLength - len(prefix)
+		left := autoNamingSpec.MaxLength - len(prefix) - len(namingTrivia.suffix)
 
 		if left <= 0 {
-			return resource.PropertyValue{}, fmt.Errorf("failed to auto-generate value for %[1]q."+
-				" Prefix: %[2]q is too large to fix max length constraint of %[3]d. Please provide a value for %[1]q",
-				sdkName, prefix, autoNamingSpec.MaxLength)
+			if len(namingTrivia.suffix) > 0 {
+				return resource.PropertyValue{}, fmt.Errorf("failed to auto-generate value for %[1]q."+
+					" Prefix: %[2]q is too large to fix max length constraint of %[3]d"+
+					" with required suffix %[4]q. Please provide a value for %[1]q",
+					sdkName, prefix, autoNamingSpec.MaxLength, namingTrivia.suffix)
+			} else {
+				return resource.PropertyValue{}, fmt.Errorf("failed to auto-generate value for %[1]q."+
+					" Prefix: %[2]q is too large to fix max length constraint of %[3]d. Please provide a value for %[1]q",
+					sdkName, prefix, autoNamingSpec.MaxLength)
+			}
```

Comment:
Kind of an aside - but it's a little bit of a shame how much we "reinvent" across providers for autonaming.  Feels like something where we could push a standard library of autonaming utilities into pulumi/pulumi or elsewhere to apply here and elswhere.

TFbridge has a somewhat rich set of utilities for autonaming that can be applied to do I believe everything done here as special cases.  And AFAICT - nothing in `FromName` there is specific to Terraform - so that pattern could be lifted out and used elsewhere potentially.

https://github.com/pulumi/pulumi-terraform-bridge/blob/master/pkg/tfbridge/names.go#L240-L260

Reply:
Agreed - it's all resource spec / attribute mappings to string manipulation which should just not care how the resources themselves were defined.  I'd love to chat about how to handle this more generally sometime - especially when it comes to provider-specific semantics.

Reply:
@lukehoban - still working on polishing up some of the code + comments in this PR, but per @mikhailshilkov 's feedback, I've made this a bit more generalized and integrated it into schema generation for this repo.  That feels a lot more generic wrt which native provider is being tinkered with and may be a potential path forward in other places.